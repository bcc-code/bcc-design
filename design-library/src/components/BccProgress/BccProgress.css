@layer components {
    /* Base */
    .bcc-progress-container {
        @apply inline-flex flex-col space-y-2;
    }

    .bcc-progress {
        @apply rounded w-full bg-neutral-200;
    }

    /* Color */
    .bcc-progress-line {
        @apply rounded bg-silver-tree-600;
    }

    /* Size */
    .bcc-progress-container .bcc-form-label-lg .bcc-form-label-optional {
        @apply text-label-base;
    }
}
