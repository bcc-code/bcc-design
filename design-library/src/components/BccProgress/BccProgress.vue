<script setup lang="ts">
import { computed } from "vue";
import BccFormLabel from "@/components/BccFormLabel/BccFormLabel.vue";

type Props = {
  value?: number;
  max?: number;
  size?: "sm" | "base" | "lg";
  progressClass?: string | object;
  barClass?: string | object;
  labelClass?: string | object;
  showValues?: boolean;
  showPercentage?: boolean;
  exceed?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  value: 0,
  max: 100,
  size: "base",
  showValues: true,
  showPercentage: true,
  exceed: true,
});

const percentage = computed(() => {
  if (props.value > props.max && !props.exceed) return 100;
  return (props.value / props.max) * 100;
});

const percentageSafe = computed(() => {
  if (props.value > props.max) return 100;
  return percentage.value;
});

const progressClass = computed(() => {
  if (props.progressClass) return props.progressClass;
  return {
    "bcc-progress": true,
    "h-1": props.size === "sm",
    "h-1.5": props.size === "base",
    "h-2": props.size === "lg",
  };
});

const barClass = computed(() => {
  if (props.barClass) return props.barClass;
  if (
    typeof progressClass.value === "object" &&
    Object.keys(progressClass.value).includes("bcc-progress")
  ) {
    const clone: any = { ...progressClass.value };
    delete clone["bcc-progress"];
    return {
      ...clone,
      "bcc-progress-bar": true,
    };
  }
  return progressClass.value;
});
</script>

<template>
  <div class="bcc-progress-container">
    <BccFormLabel
      v-if="showPercentage || showValues"
      :size="size !== 'sm' ? size : 'base'"
      :showOptionalLabel="showPercentage"
      :optionalLabel="Math.round(percentage) + '%'"
      :class="labelClass"
    >
      <span v-if="showValues">
        {{ props.value.toLocaleString() }} /
        {{ props.max.toLocaleString() }}
      </span>
    </BccFormLabel>

    <div class="bcc-progress" :class="progressClass">
      <div
        class="bcc-progress-bar h-full"
        :class="barClass"
        :style="{ width: percentageSafe + '%' }"
      />
    </div>
  </div>
</template>
