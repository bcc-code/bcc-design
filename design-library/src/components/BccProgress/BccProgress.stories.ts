import BccProgress from "./BccProgress.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

/**
 * A `progress` element supporting current and max values
 */
export default {
  title: "Forms/BccProgress",
  component: BccProgress,
  argTypes: {
    value: {
      description: "Current Value",
      control: { type: "number" },
    },
    size: {
      options: ["sm", "base", "lg"],
      control: { type: "radio" },
    },
    max: {
      description: "Max Value",
      control: { type: "number" },
    },
    exceed: {
      description: "Let or let not percentage exceed 100%",
    },
    showValues: {
      description: "Display current and max value",
    },
    showPercentage: {
      description: "Display percentage",
    },
  },
} as Meta<typeof BccProgress>;

const Template: StoryFn<typeof BccProgress> = (args) => ({
  components: { BccProgress },
  setup() {
    return { args };
  },
  template: `
    <BccProgress v-bind="args" class="w-1/4"/>
  `,
});

/**
 * Use `value` to control the progress of the progress bar. Pass max value to declare when it reaches 100% and set exceed to false if it should not exceed 100%.
 */
export const Example = Template.bind({});
Example.args = {
  value: 30145.50,
  max: 200000,
  size: "base",
  exceed: true,
  showValues: true,
  showPercentage: true,
};
Example.parameters = {
  docs: {
    source: {
      language: "html",
      code: `
<BccProgress
  value="30145.50"
  max="200000"
  size="base"
  showValues="true"
  showPercentage="true"
  exceed="true"
/>
`,
    },
  },
};

/**
 * Set the `size` prop to control the size
 */
export const Size: StoryFn<typeof BccProgress> = () => ({
  components: { BccProgress },
  template: `
    <div class="flex flex-col space-y-4">
      <BccProgress class="w-1/4" size="sm" :value="30"/>
      <BccProgress class="w-1/4" size="base" :value="75"/>
      <BccProgress class="w-1/4" size="lg" :value="22"/>      
    </div>
  `,
});

/**
 * Set the `max` value, it supports both decimals and integers
 */
export const Max: StoryFn<typeof BccProgress> = () => ({
  components: { BccProgress },
  template: `
    <div class="flex flex-col space-y-4">
      <BccProgress class="w-1/4" :value="0.435" :max="0.9"/>
      <BccProgress class="w-1/4" :value="8" :max="10"/>   
    </div>
  `,
});

/**
 * Do or do not let percentage `exceed` 100%
 */
export const exceed: StoryFn<typeof BccProgress> = () => ({
  components: { BccProgress },
  template: `
    <div class="flex flex-col space-y-4">
      <BccProgress class="w-1/4" :value="1050" :max="100"/>
      <BccProgress class="w-1/4" :value="1050" :max="100" :exceed="false"/>
    </div>
  `,
});

/**
 * Set `showValues` and / or `showPercentage` to false
 */
export const showValuesAndOrPercentage: StoryFn<typeof BccProgress> = () => ({
  components: { BccProgress },
  template: `
    <div class="flex flex-col space-y-4">
      <BccProgress class="w-1/4" :value="8" :max="10"/>
      <BccProgress class="w-1/4" :value="8" :max="10" :showValues="false"/>   
      <BccProgress class="w-1/4" :value="8" :max="10" :showPercentage="false"/>   
      <BccProgress class="w-1/4" :value="8" :max="10" :showValues="false" :showPercentage="false"/>   
    </div>
  `,
});
