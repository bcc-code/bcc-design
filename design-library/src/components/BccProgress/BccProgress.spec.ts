import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccProgress from "./BccProgress.vue";

describe("BccProgress", () => {

  it("renders without any properties", () => {
    const wrapper = mount(BccProgress);

    expect(wrapper.text()).toContain("0%");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should exceed 100%", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 110, max: 100 },
    });

    expect(wrapper.text()).toContain("110%");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should not exceed 110%", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 110, max: 100, exceed: false },
    });

    expect(wrapper.text()).toContain("100%");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render percentage", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 100, max: 100, showPercentage: true },
    });

    expect(wrapper.text()).toContain("100%");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should not render percentage", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 100, max: 100, showPercentage: false },
    });

    expect(wrapper.text()).not.toContain("100%");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render values", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 100, max: 100, showValues: true },
    });

    expect(wrapper.text()).toContain("100 / 100");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should not render values", () => {
    const wrapper = mount(BccProgress, {
      props: { value: 100, max: 100, showValues: false },
    });

    expect(wrapper.text()).not.toContain("100 / 100");
    expect(wrapper.html()).toMatchSnapshot();
  });

});
