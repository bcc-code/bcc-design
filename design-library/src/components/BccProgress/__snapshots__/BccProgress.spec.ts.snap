// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccProgress > renders without any properties 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>0 / 100</span></span><span class=\\"bcc-form-label-optional\\">0%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 0%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should exceed 100% 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>110 / 100</span></span><span class=\\"bcc-form-label-optional\\">110%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should not exceed 110% 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>110 / 100</span></span><span class=\\"bcc-form-label-optional\\">100%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should not render percentage 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>100 / 100</span></span>
    <!--v-if-->
  </label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should not render values 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span></span><span class=\\"bcc-form-label-optional\\">100%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should render percentage 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>100 / 100</span></span><span class=\\"bcc-form-label-optional\\">100%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;

exports[`BccProgress > should render values 1`] = `
"<div class=\\"bcc-progress-container\\"><label class=\\"bcc-form-label\\"><span><span>100 / 100</span></span><span class=\\"bcc-form-label-optional\\">100%</span></label>
  <div class=\\"bcc-progress h-1.5\\">
    <div class=\\"bcc-progress-line h-1.5\\" style=\\"width: 100%;\\"></div>
  </div>
</div>"
`;
